python/ray/data/dataset.py

```diff
@@ -845,15 +846,115 @@ def groupby(self, key: "GroupKeyT") -> "GroupedDataset[T]":
         from ray.data.grouped_dataset import GroupedDataset
         return GroupedDataset(self, key)
 
-    def aggregate(self, *aggs: AggregateFn) -> U:
-        """Aggregate the entire dataset as one group.
+    def agg(self, aggs: Dict[AggregateOnT, List["Aggregation"]]):
+        """Aggregate using one or more operations over the specified data
+        subsets.
+
+        Common aggregations can be given as strings (``"sum"``, ``"min"``,
+        ``"max"``, ``"mean"``, ``"std"``) while custom/customized aggregations
+        can be given as instances of :class:`~.AggregateFn`. If providing a
+        dictionary of data subsets --> aggregations, the data subset keys can
+        be column name strings for table (Arrow) datasets, and can be
+        row --> data subset callables for both table (Arrow) datasets and
+        simple datasets. See examples below.
+
+        This is a blocking operation.
+
+        Examples:
+            >>> ray.data.range(100).agg({
+            ...     lambda x: x % 3: ["min", "max"],
+            ...     lambda x: x % 2: ["mean", "std"],
+            ... })
+            >>> ray.data.from_items([
+            ...         {"A": x, "B": x**2} for x in range(100)])
+            ...     .agg({
+            ...         "A": ["mean", Std(ddof=0)],
+            ...         "B": ["min", "max"],
```

Comment:
Btw what happens if you do: `lambda x: x["A"] + x["B"]: ["min"]`? Is a reasonable column name generated for the output?

Reply:
The column names are best effort, pulled from the function name, so the user is expected to define a function with a sensible name if they care about the output column name.

Reply:
What actually happens here though? Do you end up with a column like `<anonymous lambda 0x43234>`?

Reply:
Yes, the column name inherits the opaque lambda function name, which is why users would need to define a function with a sensible name. If this best-effort output column naming won't suffice, we'll need to give the user some way to specify the output column names in the `.agg()` API, something like what I indicated here: https://github.com/ray-project/ray/pull/20090#discussion_r749931037